Tow v1.0

claxn opened the pull request:
- Added "from_parameter" parsing and linkage to process specifications (URL, directory path or list of processes)
- Enhanced user-friendliness with sub-class `OpenEONode`
- Added plotting functionality with `igraph`
- Completed documentation
- Added examples how to work with the package

m-mohr commented:
What I couldn't really figure out: How to get the (default) value for a parameter?

claxn commented:
@m-mohr 
As an example I have chosen the EVI process graph.
After you have accessed a node, e.g., via `process_graph['p2_7']` you can access its parameters/arguments via `process_graph['p2_7'].arguments`. What you get is the following:
```
{'data': [-7.5, {'from_node': 'blue_4'}],
 'ignore_nodata': <openeo_pg_parser.definitions.OpenEOParameter at 0x28b2c4d44e0>}
```
Its a dictionary, where you can access each parameter via `process_graph['p2_7'].arguments['data']` or `process_graph['p2_7'].arguments['ignore_nodata']`. 'ignore_nodata' is not given in the process graph and is therefore replaced by an instance of `OpenEOParameter`. This instance has attributes like `default_value`, `schema`, `is_optional` or `name`. So if you execute:
```
process_graph['p2_7'].arguments['ignore_nodata'].default_value
```
You will get `True`. I still need to add this to the documentation. Or would you/other reviewers prefer a different way of accessing default values/parameters? Maybe it is more consistent if every parameter is an `OpenEOParameter` and you always can access its value. What do you think?

m-mohr commented:
Thanks for the explanation. My question was not very good though, I was more thinking about the process parameter default values.

For example, what would `process_graph['dc'].arguments['bands']` return? It should be a list with "B08", "B04", "B02".

```json
{
  "parameters": [
    {
      "name": "collection-id",
      "description": "The ID of the collection to load.",
      "default": "COPERNICUS/S2",
      "schema": {
        "type": "string",
        "subtype": "collection-id",
        "pattern": "^[\\w\\-\\.~/]+$"
      }
    },
    {
      "name": "bands",
      "description": "The band names for the NIR, RED and BLUE bands in exactly this order.",
      "default": [
        "B08",
        "B04",
        "B02"
      ],
      "schema": {
        "type": "array",
        "items": {
          "type": "string",
          "subtype": "band-name"
        }
      }
    }
  ],
  "process_graph": {
    "dc": {
      "process_id": "load_collection",
      "description": "Loading the data; The order of the specified bands is important for the following reduce operation.",
      "arguments": {
        "id": {
          "from_parameter": "collection-id"
        },
        "spatial_extent": {
          "west": 16.1,
          "east": 16.6,
          "north": 48.6,
          "south": 47.2
        },
        "temporal_extent": [
          "2018-01-01",
          "2018-02-01"
        ],
        "bands": {
          "from_parameter": "bands"
        }
      },
	  "result": true
    }
  }
}
```

soxofaan commented on src/openeo_pg_parser/definitions.py:
```diff
@@ -0,0 +1,201 @@
+    @property
+    def description(self):
+        """ str : Description of the process. """
+        return self.definition['description'] if self._has_description else None
```
All these constructions like
```python
self.definition['foo'] if self._has_foo else None
```

can be simplified in Python to a simple `.get()` call , which allows to specify a default value:
```
self.definitions.get('foo', None)
```

Moreover, the default value of this default argument is `None`, so you can write just:
```
self.definitions.get('foo')
```

claxn replied:
Ah yes, I know, but I completely forgot to use it in this case... Thanks a lot!

soxofaan commented on src/openeo_pg_parser/processes/absolute.json:
```diff
@@ -0,0 +1,98 @@
+{
```
is this a copy of the openeo processes repository? 
Isn't it better to use git submodules for this?

in the openeo-python-driver we are using git submodules: here https://github.com/Open-EO/openeo-python-driver/tree/master/openeo_driver/specs/openeo-processes

m-mohr replied:
We were discussing recently to publish them as package as submodules sometimes are really annoying to use. Would that be useful in general?

soxofaan replied:
> publish them as package

You mean as a zip/tar file?
While I understand that git submodules take some to time to get used to,
I'm not sure whether separate packages simplify the setup/build procedure. With git submodules you just have to do something like `git submodule update` and you're set. While with a generic package you have to manage some public URL with versions/tags, download from that URL, make sure you have correct tools to upack it, unpack to correct location, etc. Not impossible, but a lot more tedious and error prone to do manually or to automate. Git submodules eliminate a lot of those troubles (at the expense of being harder to understand)

m-mohr replied:
No, I meant as NPM / PyPi packages so that they can used with the normal dependency management etc...

claxn replied:
I would agree with @m-mohr. If we aim to publish a version for each package manager (npm and pip at the moment), isn't it cleaner to do this without git submodules?

soxofaan replied:
ok, npm/PyPI packages would indeed be a handy replacement for git submodules indeed.

One disadvantage I can think of is that you then can only "include" one specific version of the openeo-processes. In the openeo-python-driver for example we currently both support 0.4 and 1.0 versions, which is possible with git submodules. But most other applications will only need to link to a single version I guess.

m-mohr replied:
That is indeed true and not sure whether there's an easy way to provide all versions through a package...

claxn replied:
Thanks both for your input. So whats the conclusion on this, should we stick to a semi-dynamic version of the processes (always updated with new releases of the parser) or use a git submodule? For the the time being I would prefer the first option, because this would be also easier to handle from the testing side at the moment. As soon as I have more time again, I can have a look into this and replace the process folder with a git submodule.

soxofaan commented on src/openeo_pg_parser/utils.py:
```diff
@@ -0,0 +1,198 @@
+    """
+
+    key_str = keys2str(keys)
+    return copy.deepcopy(eval('obj' + key_str))
```
this trick with `eval` and building a long indexing string looks bad practice to me
why not just use recursion? e.g.

```python
def get_obj_elem_from_keys(obj, keys):
    if len(keys) > 1:
        return get_obj_elem_from_keys(obj[keys[0]], keys[1:])
    else:
        return obj[keys[0]]
```

claxn replied:
I agree, thats much better and cleaner, thanks!

soxofaan commented on src/openeo_pg_parser/utils.py:
```diff
@@ -0,0 +1,198 @@
+    """
+
+    key_str = keys2str(keys)
+    exec('obj{}={}'.format(key_str, value))
```
same note as above: just use recursion here

claxn commented:
@m-mohr : the latest commit now includes (hopefully) a correct parsing of user defined parameters at any level. So if I parse your process graph I get by executing print(`process_graph`):
```
Node ID: dc_0 
Node Name: dc 
{'arguments': {'bands': ['B08', 'B04', 'B02'],
               'id': 'COPERNICUS/S2',
               'spatial_extent': {'east': 16.6,
                                  'north': 48.6,
                                  'south': 47.2,
                                  'west': 16.1},
               'temporal_extent': ['2018-01-01', '2018-02-01']},
 'description': 'Loading the data; The order of the specified bands is '
                'important for the following reduce operation.',
 'process_id': 'load_collection',
 'result': True}
```

m-mohr commented:
That looks good. I haven't looked at the code though.

claxn commented:
I have now implemented the suggested changes and fixed all raised issues. Please have a look at it, test it again for your back-end (if it applies) and let me know if I should change anything. Further enhancements will come for sure in the future, so this will just be the first release for v1.0, but this version should be at least working for everything what you need at the moment. Thanks!

lforesta commented:
I get the following error when specifying `process_defs` as a url:
```
ValueError: Process 'load_collection' could not be found in the list of processes.

The process is present at the given url, but it seems that the class `OpenEOProcess` interprets that as a filepath?

m-mohr commented:
So as there's still the viable option to release the processes separately, I'm still in favor to rename the package. Otherwise we'll run into conflicts.

m-mohr commented:
I'm curious how this process would be parsed: https://github.com/Open-EO/openeo-processes/pull/162
Especially wondering about whether sum/divide are actually called/executed for an a nulled / empty array.

claxn commented:
@lforesta : I have modified the remote download behaviour a bit and reverted it now for the processes. Please try again. The behaviour is now that processes are pulled from the overview endpoint (e.g., [https://earthengine.openeo.org/v1.0/processes](https://earthengine.openeo.org/v1.0/processes))and collections from each specific collections endpoint (e.g., [https://earthengine.openeo.org/v1.0/collections/WorldPop/POP](https://earthengine.openeo.org/v1.0/collections/WorldPop/POP)).
